### DISCUSSION TITLE
Rotate Personal Access Tokens

### DISCUSSION THREAD

**Discussion by StevenACoffman** [community/community#24366] [Apps, API and Webhooks]
<p>Our organization uses github service accounts, and those use personal access tokens for a variety of tasks. We have a requirement that all accounts have multifactor authentication (MFA/TOTP) enabled. We also have a requirement that all keys be rotated on a regular basis, and this includes personal access tokens.</p>
<p>Is there any way that a personal access token can either be created or regenerated <strong>via</strong> a personal access token <strong>without</strong> a password?</p>
<p>We can rotate the personal access token using the API using basic authentication, but currently we need both the password and the MFA TOTP which inhibits automation. An example <a>bash script to rotate personal access token is here</a>. ](<a href="https://gist.github.com/StevenACoffman/f0c084b428977430d2baacd0263c3563" class="inline-onebox-loading" rel="nofollow noopener">https://gist.github.com/StevenACoffman/f0c084b428977430d2baacd0263c3563</a>). )<br>
Any ideas? Thanks!</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>As far as I know, there is no way to generate new personal access tokens using only a personal access token. I have some <a href="https://github.com/lee-dohm/dotfiles/blob/master/bin/github-labels#L57-L92" rel="nofollow noopener">Ruby code</a> that I use to create a new personal access token inside scripts that handles 2FA. So the rotation step could be automated, but the TOTP code would still have to be supplied manually via a script, tool, or webpage at the time of rotation.</p>
<p>I hope that helps!</p>

**Comment by kuperiu** [comment]
<p>If you use vault you can use this <a href="https://github.com/kuperiu/vault-plugin-secrets-github/" rel="noopener nofollow ugc">https://github.com/kuperiu/vault-plugin-secrets-github/</a></p>

**Comment by HariSekhon** [comment]
GitHub could add an API endpoint to rotate a PAT like GitLab has:

https://docs.gitlab.com/ee/api/personal_access_tokens.html#rotate-a-personal-access-token

**Reply by dubsalot** [reply]
I work in Cyber Security @ my employer and very short TTL PAT generation is something we want to see. 

We would go so far as 'new token for every git operation' but would settle for having an API endpoint to mint a new token daily with a cron from the developer environment.

This would be very useful for us in a highly regulated industry with strict security policies.

**Comment by SaifurRR** [comment]
Is there an API for this?

**Comment by dmtzs** [comment]
So rotation of the PAT is not going to be available under any circumstances?

**Comment by SaifurRR** [comment]
you can generate personal access token to achieve this

**Reply by SANDS93** [reply]
ok

**Reply by antoniodellelce** [reply]
@SaifurRR as of beginning of 2026 I don't see a way to automate this, if I understand your sentence correctly, a PAT cannot manage PATs... so you cannot "achieve this"

**Reply by porobertdev** [reply]
move to gitlab XD


<img width="540" height="139" alt="image" src="https://github.com/user-attachments/assets/e73250a9-799b-44a0-b15b-44ee428c1eb9" />
<img width="867" height="396" alt="image" src="https://github.com/user-attachments/assets/c2166282-a727-40d8-9f03-f6fe582f0365" />
